File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
     List<List<Integer>> result=new ArrayList<>();
        if(root ==null)
            return result;
        boolean flag=false;
        Queue<TreeNode> q=new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            int size=q.size();
            List<Integer> level=new ArrayList<>();
            for(int i=0;i<size;i++)
            {
                TreeNode cur=q.poll();
                level.add(cur.val);
                if(cur.left!=null)
                    q.add(cur.left);
                if (cur.right!=null)
                    q.add(cur.right);
            }
            if(flag)
                Collections.reverse(level);
            result.add(level);
            flag=!flag;
        }
        
    return result;
    }
}

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
class Solution:
    def zigzagLevelOrder(self, root):
        result = []
        if not root:
            return result
        
        flag = True
        q = deque([root])
        
        while q:
            size = len(q)
            level = []
            
            for _ in range(size):
                cur = q.popleft()
                level.append(cur.val)
                
                if cur.left:
                    q.append(cur.left)
                if cur.right:
                    q.append(cur.right)
            
            
            result.append(level)
            flag = not flag
            if flag:
                level.reverse()
        
        return result
